spi/src/main/java/org/opensearch/jobscheduler/spi/JobExecutionContext.java

```diff
@@ -8,11 +8,15 @@
  */
 package org.opensearch.jobscheduler.spi;
 
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.io.stream.Writeable;
 import org.opensearch.jobscheduler.spi.utils.LockService;
 
+import java.io.IOException;
 import java.time.Instant;
 
-public class JobExecutionContext {
+public class JobExecutionContext implements Writeable {
     private final Instant expectedExecutionTime;
     private final JobDocVersion jobVersion;
     private final LockService lockService;
```

Comment:
For my own education, what is this `LockService` used for? It is not included in the `Writeable` implementation so it is not sent over transport.  It is only set (final) in the constructor and read via a getter (and could be null if it's been sent over transport).

Reply:
Yes, it is not included in the `Writeable` implementation for a couple of reasons, most of which have been documented in this issue here : https://github.com/opensearch-project/opensearch-sdk-java/issues/363. 

When a plugin is prompted to run their job, this `JobExecutionContext` is provided which includes information such as the job version, id, expected execution time, and also the `LockService` object. This is used by the plugin to invoke `aquireLock` prior to executing their job (which retrieves a lock model object from the lock service metadata index) or `releaseLock` after their job execution has been completed (which updates the lock index entry so that another job can be enabled to run). 

For extensions, we still want to provide the  `JobExecutionContext` when triggering an extension's `ScheduledJobRunner.runJob` implementation, hence why this class was modified to implement `Writeable`, however we cannot send the `LockService` over transport. To enable extensions to still acquire/release this lock, we shall expose two REST APIs on Job Scheduler to achieve the same, which is tracked as work items 5 and 6 within our meta issue for the communication mechanism.